File: app/src/main/java/moe/leer/grain/transcript/TranscriptAdapter.java
package moe.leer.grain.transcript;

import android.content.Context;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.TextView;
import androidx.annotation.ColorRes;
import androidx.annotation.NonNull;
import androidx.core.content.ContextCompat;
import androidx.recyclerview.widget.RecyclerView;
import moe.leer.grain.R;
import moe.leer.grain.model.Transcript;

import java.util.List;
import java.util.Locale;

/**
 * Created by leer on 1/15/19.
 * Copyright (c) 2017 LeeReindeer
 * https://github.com/LeeReindeer
 */
public final class TranscriptAdapter extends RecyclerView.Adapter<TranscriptAdapter.ViewHolder> {

    private final LayoutInflater inflater;
    private List<Transcript> transcriptList;
    private Context context;

    public TranscriptAdapter(Context context) {
        inflater = LayoutInflater.from(context);
        this.context = context;
    }

    public List<Transcript> getTranscriptList() {
        return transcriptList;
    }

    public void setTranscriptList(List<Transcript> transcriptList) {
        this.transcriptList = transcriptList;
        notifyDataSetChanged();
    }

    public Transcript getTranscriptAt(int pos) {
        return transcriptList.get(pos);
    }

    @NonNull
    @Override
    public ViewHolder onCreateViewHolder(@NonNull ViewGroup parent, int viewType) {
        return new ViewHolder(inflater.inflate(R.layout.transcript_item, parent, false));
    }

    @Override
    public void onBindViewHolder(@NonNull ViewHolder holder, int position) {
        Transcript transcript = transcriptList.get(position);

        int score = getFinalScore(transcript.getScore());
        if (transcript.getReScore() != null) {
            score = getFinalScore(transcript.getReScore());
        }
        holder.scoreText.setText(String.valueOf(score));
        holder.scoreText.setBackgroundColor(ContextCompat.getColor(context, getBackgroundColor(score)));

        String name = transcript.getSubjectName().length() > 10 ?
                transcript.getSubjectName().substring(0, 8) + "..." :
                transcript.getSubjectName();
        if (transcript.getReScore() != null) {
            name += " (重修)";
        }
        holder.nameText.setText(name);

        holder.gradeText.setText(String.format(Locale.CHINESE, "%.1f学分", Float.parseFloat(transcript.getSubjectFullGrade())));
        holder.propertyText.setText(transcript.getSubjectProperty());
        holder.timeText.setText(String.format(Locale.CHINESE, "%s 第%d学期", transcript.getYear(), transcript.getSemester()));
    }

    private int getFinalScore(String scoreStr) {
        int score = 0;
        char firstChar = scoreStr.charAt(0);
        if (firstChar >= '0' && firstChar <= '9') {
            score = Integer.parseInt(scoreStr);
        } else {
            switch (scoreStr) {
                case "优秀":
                    score = 95;
                    break;
                case "良好":
                    score = 85;
                    break;
                case "中等":
                    score = 75;
                    break;
                case "及格":
                    score = 65;
                    break;
                case "不及格":
                    score = 0;
                    break;
                default:
                    score = 0;
                    break;
            }
        }
        return score;
    }

    //todo 可以分过了和没过的
    private @ColorRes
    int getBackgroundColor(int score) {
        if (score >= 90) {
            return R.color.green;
        } else if (score >= 80) {
            return R.color.blue;
        } else if (score >= 70) {
            return R.color.orange;
        } else if (score >= 60) {
            return R.color.grey;
        } else {
            return R.color.pink;
        }
    }

    @Override
    public int getItemCount() {
        return transcriptList == null ? 0 : transcriptList.size();
    }

    class ViewHolder extends RecyclerView.ViewHolder {

        final TextView scoreText;
        final TextView nameText;
        final TextView gradeText;
        final TextView propertyText;
        final TextView timeText;

        ViewHolder(@NonNull View itemView) {
            super(itemView);

            scoreText = itemView.findViewById(R.id.scoreText);
            nameText = itemView.findViewById(R.id.subjectNameText);
            gradeText = itemView.findViewById(R.id.subjectGradeText);
            propertyText = itemView.findViewById(R.id.subjectPropertyText);
            timeText = itemView.findViewById(R.id.subjectTimeText);
        }
    }
}

File: app/src/main/java/moe/leer/grain/transcript/TranscriptAdapter.kt
package moe.leer.grain.transcript

import android.content.Context
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.TextView
import androidx.annotation.ColorRes
import androidx.core.content.ContextCompat
import androidx.recyclerview.widget.RecyclerView
import moe.leer.grain.Constant
import moe.leer.grain.R
import moe.leer.grain.getSP
import moe.leer.grain.model.Transcript
import java.util.*

/**
 * Created by leer on 1/15/19.
 * Copyright (c) 2017 LeeReindeer
 * https://github.com/LeeReindeer
 */
class TranscriptAdapter(private val context: Context) : RecyclerView.Adapter<TranscriptAdapter.ViewHolder>() {

    private val inflater: LayoutInflater

    var transcriptList: List<Transcript>? = null
        set(transcriptList) {
            field = transcriptList
            notifyDataSetChanged()
        }

    init {
        inflater = LayoutInflater.from(context)
    }

    fun getTranscriptAt(pos: Int): Transcript {
        return this.transcriptList!![pos]
    }

    override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): ViewHolder {
        return ViewHolder(inflater.inflate(R.layout.transcript_item, parent, false))
    }

    override fun onBindViewHolder(holder: ViewHolder, position: Int) {
        val transcript = this.transcriptList!![position]
        with(transcript) {
            with(holder) {
                var showScore = getFinalScore(score)
                if (reScore != null) {
                    showScore = getFinalScore(reScore)
                }
                scoreText.text = showScore.toString()
                scoreText.setBackgroundColor(ContextCompat.getColor(context, getBackgroundColor(showScore)))
                usualGradeText.text = usualGrade ?: context.getString(R.string.test_na)
                examGradeText.text = examGrade ?: context.getString(R.string.test_na)

                var name = if (subjectName.length > 10)
                    subjectName.substring(0, 8) + "..."
                else
                    subjectName
                if (reScore != null) {
                    name += " (重修)"
                }
                nameText.text = name

                gradeText.text =
                    String.format(Locale.CHINESE, "学分:%.1f", java.lang.Float.parseFloat(subjectFullGrade))
                propertyText.text = subjectProperty
                timeText.text = String.format(Locale.CHINESE, "%s-%d", year, semester)
            }

        }

    }

    private fun getFinalScore(scoreStr: String): Int {
        var score = 0
        val firstChar = scoreStr[0]
        if (firstChar in '0'..'9') {
            score = Integer.parseInt(scoreStr)
        } else {
            when (scoreStr) {
                "优秀" -> score = 95
                "良好" -> score = 85
                "中等" -> score = 75
                "及格" -> score = 65
                "不及格" -> score = 0
                else -> score = 0
            }
        }
        return score
    }

    @ColorRes
    private fun getBackgroundColor(score: Int): Int {
        if (context.getSP(Constant.SP_SETTING_NAME).getString("key_transcript", "0") == "1") {
            return if (score >= 60) {
                R.color.green
            } else {
                R.color.pink
            }
        }
        return if (score >= 90) {
            R.color.green
        } else if (score >= 80) {
            R.color.blue
        } else if (score >= 70) {
            R.color.orange
        } else if (score >= 60) {
            R.color.grey
        } else {
            R.color.pink
        }
    }

    override fun getItemCount(): Int {
        return if (this.transcriptList == null) 0 else this.transcriptList!!.size
    }

    class ViewHolder(itemView: View) : RecyclerView.ViewHolder(itemView) {

        val scoreText: TextView = itemView.findViewById(R.id.finalGradeText)
        val usualGradeText: TextView = itemView.findViewById(R.id.usualGradeText)
        val examGradeText: TextView = itemView.findViewById(R.id.examGradeText)
        val nameText: TextView = itemView.findViewById(R.id.subjectNameText)
        val gradeText: TextView = itemView.findViewById(R.id.subjectGradeText)
        val propertyText: TextView = itemView.findViewById(R.id.subjectPropertyText)
        val timeText: TextView = itemView.findViewById(R.id.subjectTimeText)

    }
}
